File: Languages/cpp/algorithms-dsa/leetcode/arrays-hashing/subarray_sum_equals_k.cpp
/*
 * LeetCode 560 - Subarray Sum Equals K
 * Topic: Arrays & Hashing
 * Difficulty: Medium
 *
 * Use prefix sum with hash map: count how many previous prefix sums equal (current - k).
 * Time: O(n)
 * Space: O(n)
 */
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cassert>
using namespace std;

int subarraySum(vector<int>& nums, int k) {
    unordered_map<int, int> prefixCount;
    prefixCount[0] = 1;
    int sum = 0, count = 0;
    for (int n : nums) {
        sum += n;
        if (prefixCount.count(sum - k))
            count += prefixCount[sum - k];
        prefixCount[sum]++;
    }
    return count;
}

int main() {
    vector<int> t1 = {1,1,1};
    assert(subarraySum(t1, 2) == 2);

    vector<int> t2 = {1,2,3};
    assert(subarraySum(t2, 3) == 2);

    vector<int> t3 = {1,-1,0};
    assert(subarraySum(t3, 0) == 3);

    vector<int> t4 = {3,4,7,2,-3,1,4,2};
    assert(subarraySum(t4, 7) == 4);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/arrays-hashing/subarray_sum_equals_k.py
"""
LeetCode #560 - Subarray Sum Equals K
Topic: Arrays & Hashing
Difficulty: Medium

Count subarrays with sum equal to k using prefix sum + hashmap.

Time Complexity: O(n)
Space Complexity: O(n)
"""
from collections import defaultdict


def subarray_sum(nums: list[int], k: int) -> int:
    prefix_counts: dict[int, int] = defaultdict(int)
    prefix_counts[0] = 1
    current_sum = 0
    count = 0
    for num in nums:
        current_sum += num
        count += prefix_counts[current_sum - k]
        prefix_counts[current_sum] += 1
    return count


if __name__ == "__main__":
    assert subarray_sum([1, 1, 1], 2) == 2
    assert subarray_sum([1, 2, 3], 3) == 2
    assert subarray_sum([1, -1, 0], 0) == 3
    assert subarray_sum([0, 0, 0], 0) == 6
    print("All tests passed!")
